File: java/src/main/java/io/struxel/sdk/Guard.java
package io.struxel.sdk;

import java.util.logging.Logger;

public class Guard {
  private static final Logger LOGGER = Logger.getLogger(Guard.class.getName());

  private String enforcementMode;
  private int rateLimit;

  public Guard(String apiKey) {
    this(apiKey, "realtime", 500, LicenseValidator.DEFAULT_URL, 800, null);
  }

  public Guard(String apiKey, String enforcementMode, int rateLimit, String licenseServerUrl, int timeoutMs) {
    this(apiKey, enforcementMode, rateLimit, licenseServerUrl, timeoutMs, null);
  }

  public Guard(
      String apiKey,
      String enforcementMode,
      int rateLimit,
      String licenseServerUrl,
      int timeoutMs,
      LicenseValidator validator) {
    this.enforcementMode = enforcementMode;
    this.rateLimit = rateLimit;

    LicenseValidator validatorToUse =
        validator != null ? validator : new LicenseValidator(apiKey, licenseServerUrl, timeoutMs, "1.0.0", "java", null);

    LicenseValidator.LicenseInfo info = validatorToUse.validate();
    if (info.enforcementMode != null) {
      this.enforcementMode = info.enforcementMode;
    }
    if (info.rateLimit != null) {
      this.rateLimit = info.rateLimit;
    }
  }

  public CheckResult check(String prompt, String model, String userId) {
    return evaluate(prompt);
  }

  public CheckResult checkOutput(String output, String model) {
    return evaluate(output);
  }

  public CheckResult checkOutput(String output, String model, String userId) {
    return evaluate(output);
  }

  private CheckResult evaluate(String content) {
    long startNs = System.nanoTime();
    boolean blocked = content.contains("<struxel:block>");
    double latencyMs = (System.nanoTime() - startNs) / 1_000_000.0;

    CheckResult result = new CheckResult(
        blocked,
        blocked ? "Content violates policy" : null,
        blocked ? "policy.blocked_content" : null,
        enforcementMode,
        latencyMs);

    if (!blocked) {
      return result;
    }

    if ("realtime".equals(enforcementMode)) {
      throw new PolicyViolationException("Content blocked by Struxel Guard.", result);
    }

    if ("batch".equals(enforcementMode)) {
      LOGGER.info("Struxel violation recorded: " + result.getReason());
    } else {
      LOGGER.warning("Struxel advisory warning: " + result.getReason());
    }

    return result;
  }

  public int getRateLimit() {
    return rateLimit;
  }

  public String getEnforcementMode() {
    return enforcementMode;
  }
}

File: python/struxel_sdk/guard.py
import logging
import time
from typing import Optional

from .exceptions import PolicyViolationError
from .license import LicenseValidator
from .models import CheckResult


class Guard:
    def __init__(
        self,
        api_key: str,
        enforcement_mode: str = "realtime",
        rate_limit: int = 500,
        license_server_url: str = LicenseValidator.DEFAULT_URL,
        timeout_ms: int = 800,
        validator: Optional[LicenseValidator] = None,
    ) -> None:
        self._validator = validator or LicenseValidator(
            api_key=api_key,
            license_server_url=license_server_url,
            timeout_ms=timeout_ms,
        )
        license_data = self._validator.validate()
        self.enforcement_mode = license_data.get("enforcement_mode", enforcement_mode)
        self.rate_limit = int(license_data.get("rate_limit", rate_limit))

    def check(self, prompt: str, model: str, user_id: str) -> CheckResult:
        return self._evaluate(prompt)

    def check_output(self, output: str, model: str, user_id: Optional[str] = None) -> CheckResult:
        return self._evaluate(output)

    def _evaluate(self, content: str) -> CheckResult:
        start = time.perf_counter()
        blocked = "<struxel:block>" in content
        latency_ms = (time.perf_counter() - start) * 1000

        result = CheckResult(
            blocked=blocked,
            reason="Content violates policy" if blocked else None,
            policy_id="policy.blocked_content" if blocked else None,
            enforcement_mode=self.enforcement_mode,
            latency_ms=latency_ms,
        )

        if not blocked:
            return result

        if self.enforcement_mode == "realtime":
            raise PolicyViolationError("Content blocked by Struxel Guard.", result=result)
        if self.enforcement_mode == "batch":
            logging.info("Struxel violation recorded: %s", result.reason)
        else:
            logging.warning("Struxel advisory warning: %s", result.reason)

        return result
